src/lib/Repository/ContentService.php

```diff
@@ -381,20 +382,39 @@ public function loadContentByVersionInfo(APIVersionInfo $versionInfo, array $lan
     public function loadContent(int $contentId, array $languages = null, ?int $versionNo = null, bool $useAlwaysAvailable = true): APIContent
     {
         $content = $this->internalLoadContentById($contentId, $languages, $versionNo, $useAlwaysAvailable);
-
         if (!$this->permissionResolver->canUser('content', 'read', $content)) {
             throw new UnauthorizedException('content', 'read', ['contentId' => $contentId]);
         }
         if (
             !$content->getVersionInfo()->isPublished()
             && !$this->permissionResolver->canUser('content', 'versionread', $content)
         ) {
-            throw new UnauthorizedException('content', 'versionread', ['contentId' => $contentId, 'versionNo' => $versionNo]);
+            if (!$this->isInGracePeriod($content, $this->settings['grace_period_in_seconds'], $versionNo)) {
+                throw new UnauthorizedException('content', 'versionread', ['contentId' => $contentId, 'versionNo' => $versionNo]);
+            }
         }
 
         return $content;
     }
 
+    private function isInGracePeriod(APIContent $content, int $graceSeconds, ?int $versionNo): bool
+    {
+        if ($graceSeconds <= 0 || $versionNo === null) {
+            return false;
+        }
+
+        try {
+            $lastArchivedVersionNos = $this->persistenceHandler->contentHandler()->loadVersionNoArchivedWithin(
+                $content->getId(),
+                $graceSeconds
+            );
+        } catch (APINotFoundException $e) {
+            return false;
+        }
+
+        return in_array($versionNo, $lastArchivedVersionNos);
```

Comment:
```suggestion
        return in_array($versionNo, $lastArchivedVersionNos, true);
```
Force strict comparison. If one of the databases responded with string ID's, then we should deal with it on the gateway level to ensure data type consistency. Newer DoctrineDatabase is able to do that, don't know about the legacy one.